[k8s_cloud] Add sshjump host support.

aviweit opened the pull request:
Add an initial support for ssh jump host.

By this, only single port (nodeport) is opened on a target k8s cluster where ssh connections established from skypilot user targeted to ray cluster - are tunneled through ssh jump host pod.

aviweit commented on sky/backends/backend_utils.py:
```diff
@@ -958,7 +957,6 @@ def write_cluster_config(
 
                 # Kubernetes only:
                 'skypilot_k8s_image': k8s_image,
-                'ssh_key_secret_name': ssh_key_secret_name,
```
`ssh_key_secret_name` doesn't seem to be used.

romilbhardwaj replied:
Ah thanks for catching - this is removed in the current latest k8s_cloud branch.

aviweit commented on sky/templates/kubernetes-ray.yml.j2:
```diff
@@ -17,7 +17,7 @@ provider:
 
     # Use False if running from outside of k8s cluster
-    use_internal_ips: false
+    use_internal_ips: true
```
It causes ssh itself (not its proxy command) to target internal pod ipaddress.

aviweit commented on sky/backends/backend_utils.py:
I am not sure how to use `ssh_proxy_command_config` handled [here](https://github.com/skypilot-org/skypilot/blob/e66cda15b0effa03299066414a3828679dc6592a/sky/backends/backend_utils.py#L874). Our ssh_proxy seems to be dynamic; port is dynamic NodePort.

romilbhardwaj replied:
Can we assume port assigned to nodeport doesn't change (since the service is created only once)?

aviweit replied:
Yes. The port is the same as long as the sshjump service is running.

aviweit commented:
I confirm that launching the following task against kind, works well. 

SSH interactions with ray head during provision phase, is done through sshjump host as well as user can access ray head with ssh where head is being reached via sshjump pod. 

`hello.yaml`:

```
num_nodes: 1

resources:
  memory: 2
  cpus: 1
  # Optional; if left out, pick from the available clouds.
  cloud: kubernetes
workdir: .

setup: |
  echo "Running setup."

run: |
  echo "Hello, User, how are you?"
  conda env list
```

create a local kind k8s cluster:

```
sky local up
```

ensure cluster is up: 

```
kubectl get nodes | grep skypilot-control-plane
```

launch the task

```
sky launch hello.yaml
```

romilbhardwaj commented on Dockerfile_k8s_sshjump:
```diff
@@ -0,0 +1,24 @@
+FROM continuumio/miniconda3:22.11.1
+
+RUN apt update -y && \
+    apt install sudo vim openssh-server -y && \
+    rm -rf /var/lib/apt/lists/*
+
+# Setup SSH and generate hostkeys
+RUN mkdir -p /var/run/sshd && \
+    sed -i 's/PermitRootLogin prohibit-password/PermitRootLogin yes/' /etc/ssh/sshd_config && \
+    sed 's@session\s*required\s*pam_loginuid.so@session optional pam_loginuid.so@g' -i /etc/pam.d/sshd && \
+    cd /etc/ssh/ && \
+    ssh-keygen -A
+
+RUN useradd -m -s /bin/bash sky && \
+    echo "sky ALL=(ALL) NOPASSWD:ALL" >> /etc/sudoers && \
+    echo 'Defaults        secure_path="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"' > /etc/sudoers.d/sky
+
+# Switch to sky user
+USER sky
+
+# Add /home/sky/.local/bin/ to PATH
+RUN echo 'export PATH="$PATH:$HOME/.local/bin"' >> ~/.bashrc
+
+WORKDIR /home/sky
```
Curious - what do you think about reusing the image generated from Dockerfile_k8s (`us-central1-docker.pkg.dev/skypilot-375900/skypilotk8s/skypilot:latest`) instead of creating a new one?

Pros: Reduces image maintenance burden for us, may be faster to pull since cluster would likely already have the main `skypilot:latest` image for running tasks
Cons: Dockerfile_k8s is quite bloated because it installs SkyPilot and its dependencies.

Might be worth looking at image sizes for both before taking a call.

aviweit replied:
@romilbhardwaj , done.

romilbhardwaj commented on sky/authentication.py:
```diff
@@ -10,6 +10,8 @@
 from typing import Any, Dict, Tuple
 import uuid
 
+from urllib.parse import urlparse
```
nit: since urllib is standard library, perhaps should be added at L11 before `import uuid`.

aviweit replied:
Hi @romilbhardwaj , I removed `import urlparse`, as it is now being used by `clouds.Kubernetes.get_external_ip()` at `setup_kubernetes_authentication`.

romilbhardwaj commented on sky/authentication.py:
```diff
@@ -488,4 +490,112 @@ def setup_kubernetes_authentication(config: Dict[str, Any]) -> Dict[str, Any]:
+    sshjump_name = clouds.Kubernetes.SKY_SSH_JUMP_NAME
+    sshjump_image =  clouds.Kubernetes.SSH_JUMP_IMAGE
+
+    def _to_sshjump_pod_spec(name, image, secret):
```
Can we define this pod spec as a jinja template (e.g., `templates/k8s_ssh_jump.j2`) that is filled in with `name`, `image` and `secret` through a function call? Same for the service spec below. Might be able to put them in the same file too.

aviweit replied:
Yes, I pushed the changes.

romilbhardwaj reviewed (commented):
Thanks @aviweit! This is great. I skimmed through the code and left some comments. Haven't tried running yet. Will take another pass soon.

Can you also merge the latest changes from `k8s_cloud` branch? It has a few fixes, and some changes in how ssh auth works (especially after #2119), so would be good to verify this still works.

aviweit commented:
Hi @romilbhardwaj , I have merged the PR with `k8s_cloud` branch and tested again (with the additional commits) - with the task that is mentioned at https://github.com/skypilot-org/skypilot/pull/2201#issuecomment-1629020177.

It worked well with skypilot installed on Ubuntu 20.04 host - against an external kubernetes cluster (deployed via kubeadm), and a one deployed locally (kind).

romilbhardwaj commented on sky/authentication.py:
```diff
@@ -401,4 +404,47 @@ def setup_kubernetes_authentication(config: Dict[str, Any]) -> Dict[str, Any]:
+    content = json.loads(_content)
+
+    try:
+        kubernetes.core_api().create_namespaced_pod('default', content['pod_spec'])
```
We need to think a bit about the lifecycle of this jump pod. 

* **Creation** - Creation of this jump ssh pod happens when the user creates their first SkyPilot cluster in the Kubernetes cluster. 
* **Run** - The jump SSH pod is kept alive as long as there is a SkyPilot cluster running (i.e., there exists a Kubernetes pod with label `parent: skypilot` in the same namespace)
* **Termination** - If there hasn't been a pod with label `parent: skypilot` in the same namespace since the past n minutes (say, 10 minutes), this jump pod should terminate itself.

We can do this by running a script like this in the jump ssh pod as the `command` for the pod:

```python
import os
import datetime
import pytz
import time
from kubernetes import client, config

# Load kube config
config.load_incluster_config()

v1 = client.CoreV1Api()

# Get the current namespace from the pod service account
with open("/var/run/secrets/kubernetes.io/serviceaccount/namespace", "r") as f:
    current_namespace = f.read()

# Set the time delta for checking last active pods
time_delta = datetime.timedelta(minutes=10)

# Set delay for each retry
retry_delay = 60  # In seconds

while True:
    # Get the current time
    now = datetime.datetime.now(pytz.UTC)

    found = False
    # List the pods in the current namespace
    ret = v1.list_namespaced_pod(current_namespace)
    for i in ret.items:
        if i.metadata.labels and 'parent' in i.metadata.labels and i.metadata.labels['parent'] == 'skypilot':
            # Calculate the elapsed time since the pod was last active
            elapsed_time = now - i.metadata.creation_timestamp
            # If the pod was active in the last 10 minutes, set found to True
            if elapsed_time < time_delta:
                found = True
                break

    # If no active pods were found with the specified label, exit the script
    if not found:
        print("No active pods found with label 'parent: skypilot' in the past 10 minutes. Exiting...")
        exit(1)

    # If pods were found, sleep for the specified delay and then retry
    print(f"Active pods found with label 'parent: skypilot'. Retrying in {retry_delay} seconds...")
    time.sleep(retry_delay)
```

What do you think?

romilbhardwaj commented on sky/templates/kubernetes-sshjump.j2:
```diff
@@ -0,0 +1,74 @@
+{
```
Can we have this file as a YAML instead of a json? Trying to have something that can also be easily used with `kubectl apply -f ` for debugging/testing (of course, after replacing template variables).

romilbhardwaj reviewed (commented):
Thanks @aviweit! We probably also need to handle the lifecycle of the jump pod - left some comments.

aviweit commented:
Hi @romilbhardwaj

Thanks a lot for the comments. I updated the code to reuse Dockerfile_k8s and updated the template format to yaml. I re-tested the recent changes in our environment and it seems to work well.

Your suggestion on sshjump lifecycle seems to be fine for me. There seems to be additional work such as defining RBAC for the pod to list other pods. I tested it and got that failure.

I would like to ask whether it would be ok to merge this PR and to open another PR to handle sshjump lifecycle? Other PRs can be probably opened to handle additional enhancements as needed.
What do you think?

Thanks again.

romilbhardwaj commented:
@aviweit - this is great! Tried it out - works nicely.

> There seems to be additional work such as defining RBAC for the pod to list other pods. I tested it and got that failure.

Is there some way to upload kubeconfig credentials so that jump pod can use those to authenticate with the control plane and list pods? 

Asking because I think lifecycle management of the jump host is important to get right before we ship this. We do not want to leave an orphan ssh jump pod running on users' clusters after they are done using SkyPilot. If we can figure out a quick solution to the above problem, it'll be great to have in this PR.

If you think it will be an involved process, then it'll be good to have a draft solution verifying this kind of lifecycle management is possible. Once we can verify it can be done, then we can merge this PR and create another one to address it. 

What do you think? Thanks again for adding this!

romilbhardwaj commented:
@aviweit - I've made some updates to `k8s_cloud` branch. We use the `autoscaler` service account which has full permissions over the namespace. Is it possible to use the same `autoscaler` service account with this jump pod and use it to manage the lifecycle of the pod?

aviweit commented:
Hi @romilbhardwaj , I merged `k8s_cloud` changes into this PR and added lifecycle management of the jump host. It seems to work well in my environment.

I would like to ask whether this PR can be merged and if needed, additional PRs will be opened. This will save us time on keeping this PR up to date in respect to upstream changes.

@romilbhardwaj , I think that you will need to re-build and push Dockerfile_k8s because of adding `sshjump-lcm.py`.

Thanks again.

aviweit commented:
Hi @romilbhardwaj , I have updated the code to create sshjump related resources in the correct namespace (and a small update to `if` statement in lcm script).

Thanks a lot.

aviweit commented:
@romilbhardwaj , I added sshjump lcm to monitor only ray pods it is responsible for.
The changes are currently under sub-branch `k8s_cloud-sshjumphost2-multiuser`: https://github.com/aviweit/skypilot/tree/k8s_cloud-sshjumphost2-multiuser
commit: https://github.com/aviweit/skypilot/commit/21879c58c706476e0a2a590cd1a70354138a9b17.

If it is ok, it can be merged to this PR. Thanks.

Michaelvll commented:
We should probably re-open this PR by changing the target branch. The original k8s_cloud branch has been deleted as it was merged to the master.

romilbhardwaj commented:
Hey @aviweit ! This PR got closed when we merged https://github.com/skypilot-org/skypilot/pull/2096 into master.

For our next sprint, we'll be switching to [k8s_cloud_beta1 branch](https://github.com/skypilot-org/skypilot/tree/k8s_cloud_beta1). Can you re-open this PR on that branch?
